### ISSUE TITLE
Using expressions for dynamic content and custom functions in .moor files

### ISSUE THREAD

**Issue by laurensfischer** [simolus3/moor#1125]
Hey, 

first of all thank you for this awesome library!
We use it in our project (version ^3.4.0, because we couldn't migrate to null safety yet) and have written some complex SQL queries in .moor files.

To avoid duplicated queries, we starting using expressions like this
```
SELECT CAST($descriptionField AS TEXT) as description FROM table
WHERE
      $predicate AND $otherPredicate
GROUP BY
      $groupFields
ORDER BY $order;
```
where descriptionField is something like `CustomExpression("table.description")`.

This works just fine, but following code does not work:
`SELECT CAST($field AS TEXT) as description FROM table WHERE ? == 5 AND table.id IN ?;`
Code gets generated, but it crashes:
`Unhandled Exception: "Invalid argument (parameters): Expected 1 parameters, got 2: Instance(length:2) of '_GrowableList'"`

It works if we use it as predicate:
`SELECT id FROM table WHERE ? == 5 AND table.id IN ? AND $predicate;`

Looking at the generated code, we see that the first example produces a wrong $arrayStartIndex:
```
workingExample(int var1, List<String> var2, Expression<bool> predicate) {
    var $arrayStartIndex = 2;
    
errorExample(Expression field, int var1, List<String> var2) {
    var $arrayStartIndex = 1;
```

Is it possible to use expressions in .moor files for these scenarios?

We ran into this problem when we wanted to use custom sql functions (similar to moor_ffi_functions), which are not recognized if written in a .moor file. Is there any other way to use dynamic field names and custom functions in .moor files?

**Comment by simolus3** [comment]
> Looking at the generated code, we see that the first example produces a wrong $arrayStartIndex:

That looks like a generator bug, thanks for the report!

> we wanted to use custom sql functions (similar to moor_ffi_functions), which are not recognized if written in a .moor file

Unfortunately this is not possible without forking `moor_generator`. The `sqlparser` library (which is part of this project as well) does support custom modules that can contribute functions, but we can't inject custom Dart code into a builder. If you're interested in adding generator support for your functions, you can take a look at [this module](https://github.com/simolus3/moor/blob/master/moor_generator/lib/src/analyzer/moor/moor_ffi_extension.dart) as an example. You can add modules into the generator [here](https://github.com/simolus3/moor/blob/8f069c86f42a5b0d40719773a2da14968b6409da/moor_generator/lib/src/analyzer/session.dart#L41-L50).

**Comment by simolus3** [comment]
This was a general bug in the generator that happened if a Dart placeholder was used before an array variable. I've fixed this in 133b7c652f000654909897d421ec0f93bac9d879 (which is not yet released).

> We use it in our project (version ^3.4.0, because we couldn't migrate to null safety yet)

FWIW, you can use null safe dependencies without your app being null safe as well, this is gracefully handled by most tools. `moor_generator` 4 and above will happily generate non-nnbd code if you haven't migrated yet.
